### PR TITLE
Api request. Location permission

### PR DESCRIPTION
**Pull request by SemenciucCosmin**
FrogWeatherApiService.kt created. With that the app can now make api requests to get the forecast. Created data classes for Json objects. Implemented location permission request. The app now updates the location at startup if the last location is older than 30 minutes (when saving a new location the time difference between the old and the new location is checked in the SettingsViewModel.kt).

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -10,7 +10,7 @@ android {
 
     defaultConfig {
         applicationId "com.example.frogweather"
-        minSdk 21
+        minSdk 23
```
Comment (andreiprotiuc) [inline app/build.gradle]: why did you change from 21 to 23?

File: app/src/main/java/com/example/frogweather/MainActivity.kt [hunk 2]
```diff
@@ -1,16 +1,103 @@
...
+    }
+
+    private fun showRequestPermissionDialog() {
+        val alertDialog = AlertDialog.Builder(this, R.style.SettingsDialogTheme)
+        alertDialog.setTitle(getString(R.string.lbl_location_settings_dialog_title))
+        alertDialog.setMessage(getString(R.string.msg_location_permisson_dialog))
+        alertDialog.setNegativeButton(getString(R.string.lbl_cancel_button)) { _, _ -> }
+        alertDialog.setPositiveButton(getString(R.string.lbl_settings_button)) { _, _ ->
+            val intent = Intent()
+            intent.action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/MainActivity.kt]: action can be added to the constructor

File: app/src/main/java/com/example/frogweather/MainActivity.kt [hunk 3]
```diff
@@ -1,16 +1,103 @@
...
+            val uri = Uri.fromParts("package", packageName, null)
+            intent.data = uri
+            startActivity(intent)
+        }
+        val alert = alertDialog.create()
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/MainActivity.kt]: No need for variable here, use alertDialog.create().show()

File: app/src/main/java/com/example/frogweather/data/City.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package com.example.frogweather.data
+
+import com.squareup.moshi.Json
+
+data class City(
+    val id: Int = 0,
+    val name: String = "",
+    @Json(name = "coord")
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/data/City.kt:10]: Please add @Json for all attributes, due to obfuscation, we can discuss later

File: app/src/main/java/com/example/frogweather/data/Constants.kt [hunk 5]
```diff
@@ -21,4 +21,11 @@ const val WIND_DIRECTION_NO_INDICATION = "no indication"
 const val WIND_DIRECTION_ARROWS = "arrows"
 const val WIND_DIRECTION_ABBREVIATIONS = "abbreviations"
 const val NOTIFICATION_TYPE_DEFAULT = "default"
-const val NOTIFICATION_TYPE_SIMPLE = "simple"
\ No newline at end of file
+const val NOTIFICATION_TYPE_SIMPLE = "simple"
+const val WEATHER_API_BASE_URL = "https://api.openweathermap.org"
+const val WEATHER_API_PATH_URL = "/data/2.5/forecast?units=metric"
+const val API_KEY = "721661b090391eed20fcca7c71da55ae"
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/data/Constants.kt]: Not blocking: API key can be moved to the gradle, just a note

File: app/src/main/java/com/example/frogweather/data/FrogWeatherApiService.kt [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    .add(KotlinJsonAdapterFactory())
+    .build()
+
+private val retrofit = Retrofit.Builder()
+    .addConverterFactory(MoshiConverterFactory.create(moshi))
+    .baseUrl(WEATHER_API_BASE_URL)
+    .build()
+
+interface FrogWeatherApiService {
+    @GET(WEATHER_API_PATH_URL)
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/data/FrogWeatherApiService.kt]: You can leave here hardcoded value, it is easy to see when we have a lot of calls and files

File: app/src/main/java/com/example/frogweather/data/Main.kt [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package com.example.frogweather.data
+
+import com.squareup.moshi.Json
+
+data class Main(
+    val temp: Double = 0.0,
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/data/Main.kt]: Same here, and also the name of the object could be different, don't understand what main means

File: app/src/main/java/com/example/frogweather/data/MyLocation.kt [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.example.frogweather.data
+
+data class MyLocation(
+    val millis: Long,
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/data/MyLocation.kt:4]: Same here

File: app/src/main/java/com/example/frogweather/data/Sys.kt [hunk 9]
```diff
@@ -0,0 +1,5 @@
+package com.example.frogweather.data
+
+data class Sys(
+    val pod: String = ""
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/data/Sys.kt:7]: Same here

File: app/src/main/java/com/example/frogweather/data/Weather.kt [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package com.example.frogweather.data
+
+data class Weather(
+    val id: Int = 0,
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/data/Weather.kt:7]: Same here

File: app/src/main/java/com/example/frogweather/ui/NetworkViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+import com.example.frogweather.data.NetworkRepository
+import com.example.frogweather.data.Forecast
+import kotlinx.coroutines.launch
+
+class NetworkViewModel : ViewModel() {
+    private val networkRepository = NetworkRepository()
+
+    fun getForecast(latitude: Double, longitude: Double, millis: Long): LiveData<Forecast> {
+        val forecast = MutableLiveData<Forecast>()
+        viewModelScope.launch { forecast.value = networkRepository.getForecast(latitude, longitude, millis) }
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/ui/NetworkViewModel.kt]: I think here you will have a crash when noInternet connection, please check

File: app/src/main/java/com/example/frogweather/ui/OneDayFragment.kt [hunk 12]
```diff
@@ -5,15 +5,35 @@ import androidx.fragment.app.Fragment
...
+
         return binding.root
     }
+
+
+    companion object {
+        fun newInstance(millis: Long) : Fragment {
+            return OneDayFragment().apply {
+                arguments = Bundle().apply {
+                    putLong("key", millis)
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/ui/OneDayFragment.kt]: Use constant and also please rename to the suggested key here, something like keyMillis

File: app/src/main/java/com/example/frogweather/ui/SettingsViewModel.kt [hunk 13]
```diff
@@ -8,15 +8,34 @@ import androidx.lifecycle.ViewModel
...
     private val settingsRepository = SettingsRepository(SettingsDataSource(application))
 
-    fun getSettings(): LiveData<Settings> = settingsRepository.getSettings().asLiveData()
+    fun getSettings(): LiveData<Settings> {
+        return settingsRepository.getSettings().asLiveData()
+    }
+
+    fun getLocation(): LiveData<MyLocation> = settingsRepository.getLocation().asLiveData()
+
+
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/ui/SettingsViewModel.kt:37]: useless row here

File: app/src/main/java/com/example/frogweather/ui/SettingsViewModel.kt [hunk 14]
```diff
@@ -8,15 +8,34 @@ import androidx.lifecycle.ViewModel
...
+    fun saveLocation(location: MyLocation, context: Context) {
+        viewModelScope.launch {
+            settingsRepository.getLocation().collect { oldLocation ->
+                val minutes = (location.millis - oldLocation.millis) / MINUTES_DIVIDER
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/ui/SettingsViewModel.kt]: What you are trying to do here?

**Review by andreiprotiuc — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/frogweather/MainActivity.kt [hunk 15]
```diff
@@ -52,19 +53,16 @@ class MainActivity : AppCompatActivity() {
...
-            ContextCompat.checkSelfPermission(
-                this,
-                Manifest.permission.ACCESS_COARSE_LOCATION
-            ) == PackageManager.PERMISSION_GRANTED -> {
+            ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED -> {
                 updateLocation()
             }
-            shouldShowRequestPermissionRationale(Manifest.permission.ACCESS_COARSE_LOCATION) -> {
-                showRequestPermissionDialog()
+            Build.VERSION.SDK_INT > Build.VERSION_CODES.M -> {
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/MainActivity.kt]: >=

File: app/src/main/java/com/example/frogweather/data/FrogWeatherApiService.kt [hunk 16]
```diff
@@ -17,7 +17,7 @@ private val retrofit = Retrofit.Builder()
     .build()
 
 interface FrogWeatherApiService {
-    @GET(WEATHER_API_PATH_URL)
+    @GET("/data/2.5/forecast?units=metric")
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/data/FrogWeatherApiService.kt]: use as param, example:
suspend fun getForecast(@Query("lat") lat: Double, @Query("lon") lon: Double, @Query("dt") dt: Long, @Query("appid") appid: String, @Query("units") units = "metric"): Forecast

**Review by andreiprotiuc — CHANGES_REQUESTED** [review]

File: app/build.gradle [hunk 17]
```diff
@@ -22,6 +22,14 @@ android {
         release {
             minifyEnabled false
             proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
+
+            buildConfigField("String", "API_KEY", "\"721661b090391eed20fcca7c71da55ae\"")
+            resValue("string", "api_key", "721661b090391eed20fcca7c71da55ae")
```
Comment (andreiprotiuc) [inline app/build.gradle]: nu e duplicat aici ?

File: app/src/main/java/com/example/frogweather/data/FrogWeatherApiService.kt [hunk 18]
```diff
@@ -17,10 +19,17 @@ private val retrofit = Retrofit.Builder()
     .build()
 
 interface FrogWeatherApiService {
-    @GET("/data/2.5/forecast?units=metric")
-    suspend fun getForecast(@Query("lat") lat: Double, @Query("lon") lon: Double, @Query("dt") dt: Long, @Query("appid") appid: String): Forecast
+    @GET("/data/2.5/forecast?")
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/data/FrogWeatherApiService.kt]: no need for "?" here

File: app/src/main/java/com/example/frogweather/data/NetworkRepository.kt [hunk 19]
```diff
@@ -1,7 +1,38 @@
...
 
+import java.io.IOException
+import retrofit2.Response
+
 class NetworkRepository {
     suspend fun getForecast(latitude: Double, longitude: Double, millis: Long): Forecast {
-        return FrogWeatherApi.retrofitService.getForecast(latitude, longitude, millis, API_KEY)
+        return when (val result = safeApiCall(call = { getForecastCall(latitude, longitude, millis) }, errorMessage = "Exception occurred")) {
+            is CallResult.Success -> result.data
+            is CallResult.Error -> throw result.exception
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/data/NetworkRepository.kt]: don't need anymore to throw forward

File: app/src/main/java/com/example/frogweather/data/NetworkRepository.kt [hunk 20]
```diff
@@ -1,7 +1,38 @@
...
+        }
+    }
+
+    private suspend fun getForecastCall(latitude: Double, longitude: Double, millis: Long): CallResult<Forecast> {
+        val response: Response<Forecast>
+        try {
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/data/NetworkRepository.kt]: no need for try catch

**Review by andreiprotiuc — CHANGES_REQUESTED** [review]

File: app/build.gradle (hunk 1 above)
Comment (SemenciucCosmin) [reply app/build.gradle]: done

File: app/src/main/java/com/example/frogweather/MainActivity.kt (hunk 2 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/MainActivity.kt]: done

File: app/src/main/java/com/example/frogweather/MainActivity.kt (hunk 3 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/MainActivity.kt]: done

File: app/src/main/java/com/example/frogweather/data/City.kt (hunk 4 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/data/City.kt:10]: done

File: app/src/main/java/com/example/frogweather/data/Constants.kt (hunk 5 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/data/Constants.kt]: done

File: app/src/main/java/com/example/frogweather/data/Weather.kt (hunk 10 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/data/Weather.kt:7]: done

File: app/src/main/java/com/example/frogweather/ui/OneDayFragment.kt (hunk 12 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/ui/OneDayFragment.kt]: done

File: app/src/main/java/com/example/frogweather/ui/SettingsViewModel.kt (hunk 13 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/ui/SettingsViewModel.kt:37]: done

File: app/src/main/java/com/example/frogweather/MainActivity.kt (hunk 15 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/MainActivity.kt]: done

File: app/src/main/java/com/example/frogweather/data/FrogWeatherApiService.kt (hunk 16 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/data/FrogWeatherApiService.kt]: done

File: app/build.gradle (hunk 17 above)
Comment (SemenciucCosmin) [reply app/build.gradle]: done

File: app/src/main/java/com/example/frogweather/data/FrogWeatherApiService.kt (hunk 18 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/data/FrogWeatherApiService.kt]: done

File: app/src/main/java/com/example/frogweather/data/NetworkRepository.kt (hunk 19 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/data/NetworkRepository.kt]: done

File: app/src/main/java/com/example/frogweather/data/NetworkRepository.kt (hunk 20 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/data/NetworkRepository.kt]: done

File: app/src/main/java/com/example/frogweather/ui/NetworkViewModel.kt (hunk 11 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/ui/NetworkViewModel.kt]: yes i have  a crash. it is done

File: app/src/main/java/com/example/frogweather/ui/SettingsViewModel.kt (hunk 14 above)
Comment (SemenciucCosmin) [reply app/src/main/java/com/example/frogweather/ui/SettingsViewModel.kt]: i check how many minutes have passed since the last location update. if it is more than 30 minutes then i will update the location again. now i moved this check in the MainActivity. checking it like this didn't prevent me from requesting the location at every startup when it was unnecessary. now in the Main activity if the last location is not older than 30 minutes i won't request a new location. is it ok?

File: app/src/main/java/com/example/frogweather/MainActivity.kt [hunk 21]
```diff
@@ -80,6 +85,14 @@ class MainActivity : AppCompatActivity() {
         alertDialog.create().show()
     }
 
+    private fun shouldUpdateLocation(): Boolean {
```
Comment (andreiprotiuc) [inline app/src/main/java/com/example/frogweather/MainActivity.kt]: I don't understand how this method helps you, because the shouldUpdate will always return for the first time false, and also you still listening for the observer, but you are not doing anything with it.

**Review by andreiprotiuc — CHANGES_REQUESTED** [review]
